spec/predicates/scai.md

```diff
@@ -0,0 +1,271 @@
+# Predicate type: Software Supply Chain Attribute Integrity (SCAI)
+
+Type URI: scai/attribute-assertion/v0.1
+
+Version: 0.1
+
+Author: Marcela Melara <marcela.melara@intel.com>
+
+## Purpose
+
+The Software Supply Chain Attribute Integrity, or SCAI (pronounced "sky"), specification proposes a data
+format for capturing functional attribute and integrity information about software artifacts and their supply
+chain. SCAI data can be associated with executable binaries, statically- or dynamically-linked libraries,
+software packages, container images, software toolchains, and compute environments.
+
+Existing supply chain data formats do not capture any information about the security functionality or
+behavior of the resulting software artifact, nor do they provide sufficient evidence to support any claims of
+integrity of the supply chain processes they describe. The SCAI data format is designed to bridge this gap.
+
+For more details, see the [SCAI v0.1 specification document](https://arxiv.org/pdf/2210.05813.pdf)
+
+## Prerequisites
+
+In addition to the in-toto attestation framework,
+SCAI assumes that implementers have appropriate processes and tooling in place for
+capturing other types of software supply chain metadata, which can be extended to add support for SCAI.
+
+## Model
+
+SCAI metadata, referred to as **Attribute Assertions**, comprises a set of statements that describe functional
+attributes of a software artifact and its supply chain, capable of covering the full software stack of the
+toolchain that produced the artifact down to the hardware platform. SCAI Attribute Assertions include
+information about the conditions under which certain functional attributes arise, as well as (authenticated)
+evidence for the asserted attributes.
+
+SCAI is intended to be implemented as part of an existing software supply chain attestation
+framework by software development tools or services (e.g., builders, CI/CD pipelines, software analysis tools)
+seeking to capture more granular information about the attributes and behavior of the software artifacts they
+produce. As such, we envision SCAI metadata being explicitly bound to, or embedded within, other metadata
+objects.
+
+## Schema
+
+```
+{
+    "predicateType": "scai/attribute-assertion/v0.1",
+    "predicate": {
+        "attributes": [{
+            "attribute": "<ATTRIBUTE>",
+            "target": { // optional
+                "name": "<NAME>",
+                "digest": { "<ALGORITHM>": "<HEX VALUE>", ... },
+                "locationURI": "<URI>",
+                "objectType": "<OBJECT TYPE>" // optional
+            },
+            "conditions": { /*object */ }, // optional
+            "evidence": { /* object */ } // optional
+        }]
```

Comment:
Thinking through this, we would want to include timestamps if there is an attribute that is only valid for a period of time.

Reply:
So far, I've been thinking of attributes being quite static, but this is a really interesting point. Do you have a specific example in mind where an attribute might expire?

Reply:
I was thinking through this for GUAC and wanted to use SCAI to certify an artifact has been scanned (contains vulns or not). The attestation being valid for a period of time before it expires and needs to be re-ceritified (re-checked for vulnerabilities).

```
Certifier: The certifier creates attestations on the knowledge graph that helps translate queries about negative statements into queries on positive statements by certification. For example, instead of asking “Does artifact A have vulnerabilities?” to “Has artifact A been certified by a vulnerability scanner?”, the certifier takes in documents like OSV, VEX, bad actors and periodically (or via trigger), queries the graph to determine answers to these questions and does a certification in the knowledge graph.
```

Reply:
An example that I just made up and testing with:
```
{
  "_type": "https://in-toto.io/Statement/v0.1",
  "subject": [
      {
        "name": "git://github.com/kubernetes/kubernetes", 
        "digest": {"sha1": "5835544ca568b757a8ecae5c153f317e5736700e"}
      }
  ],
  "predicateType": "http://in-toto.io/attestation/scai/attribute-assertion/v0.1",
  "predicate": {
      "producer": {
        "type": "guac",
        "id": "guecsec/guac"
      },
      "attributes": [{
        "attribute": "scanned",
        "evidence": {
            "scanner": {
                "type": "OSV",
                "id": "osv.dev"
            },
            "results": [
              {
                "OSVID": "GHSA-jfh8-c2jp-5v3q"
              }]
        },
        "date": "2022-10-03T12:00:00Z"
      }]
  }
}
```

Reply:
I see, thanks for this example. This kind of certification process is definitely something we want to be able to support. So if I understood correctly, it's not necessarily that a certifier's attestation expires after X amount of time. Rather the problem is, how do we indicate that one attestation is no longer valid because it is replaced or superseded by a more recent one?

Reply:
Well the attestation does not need to indicate that its valid or not. That is up to the system/user to decide. Having a timestamp for when it was generated, allows GUAC or a user to determine when it was last certified as being vulnerable or not.

Reply:
Got it! Is the object in the `evidence field` something that is auto-generated by the scanner? I do wonder, since it's scan specific, if the `date` field couldn't be part of the evidence itself. But, I can also imagine there are other use cases where having a timestamp at the predicate-level could be useful, so it's definitely a field we could make optional in the SCAI predicate.

Reply:
Agreed. Updated the above example to move the date into evidence.

Reply:
Looks good, thanks! I'd definitely like to hear other's thoughts or use cases for having an optional date or timestamp field at the predicate level, too.

Reply:
Adding a timestamp makes sense to me.  The one tweak I'd suggest is being more specific about the name.  `date` could mean a number things, instead should it be named something like `scannedOn`, `collectedOn`, `generatedOn`.  This is [discussed a bit here](https://github.com/in-toto/attestation/tree/main/spec#predicate-conventions:~:text=Timestamps%20SHOULD%20use%20RFC%203339%20syntax%20with%20timezone%20%22Z%22%20and%20SHOULD%20clarify%20the%20meaning%20of%20the%20timestamp.%20For%20example%2C%20a%20field%20named%20timestamp%20is%20too%20ambiguous%3B%20a%20better%20name%20would%20be%20builtAt%20or%20allowedAt%20or%20scannedAt.).  Thoughts?

Reply:
Agreed. `scannedOn` makes more sense in this case and follows the convention.

Reply:
For a predicate-level field, I think something like `generatedOn` makes sense to keep it more generally applicable to non-scanning use cases.

Reply:
We should add a `generatedOn` for this release and change it as necessary for later releases? What do you think?